Extract type cast behaviour

zluiten opened the pull request:
This is follow up on zfcampus/zf-doctrine-querybuilder#57
<!--
Fill in the relevant information below to help triage your issue.
Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch
-->

|    Q        |   A
|------------ | ------
| Bugfix      | no
| BC Break    | no
| New Feature | yes
| RFC         | no

### Description

By extracting the type casting to its own class it can be easily replaced when needed. Preferring composition over inheritance etc.

It becomes better testable as well.

I kept BC by proxying to the TypeCaster and instantiating the TypeCaster in the AbstractFilter::getTypeCaster() method when not set in case FilterManager::filter() method is overridden.

zluiten committed:
Extract type cast behaviour

Signed-off-by: Zacharias Luiten <zach@xzachly.com>

zluiten commented:
Any idea what's needed to get this in?

weierophinney commented:
@netiul I've rebased against the new "1.9.x" branch; you'll need to pull from your own remote to get your local checkout up-to-date.

However, tests are failing.

I tried a number of different approaches to fix them, but it seems like there's something fundamentally broken in the testing suite or with the existing dependencies, and after a few hours trying to figure it all out, I need to leave that to others to resolve. Hopefully, if they get resolved, your tests will pass again, and we can merge and release.

zluiten commented:
@weierophinney Thanks for the work you do and have done! About the tests, it looks like they passed on travis?

weierophinney commented:
You're right! They hadn't been when I was testing before, but they clearly do now! The only issue is CS, which is not your problem to fix in this case (it's a line-length issue with our copyright lines). I'll get that fixed shortly.

weierophinney commented on src/Filter/ODM/AbstractFilter.php:
```diff
@@ -8,74 +8,39 @@
+
+    public function __construct($params)
     {
-        if (! isset($metadata->fieldMappings[$field])) {
-            return $value;
+        if (isset($params[1])) {
+            $this->setTypeCaster($params[1]);
         }
+    }
```
I don't like this; it makes understanding the positions of arguments paramount, which gets tricky quickly.

In looking at when this is constructed, it's from calling a plugin manager's `get()` method with the optional second argument, `$options`.

The array of options passed from a plugin manager was originally conceived as an _associative array_. That would be one way to fix this here: have the filter manager pass an associative array with named elements, which would allow this to become:

```php
public function __construct(array $params)
{
        $this->setTypeCaster($params['type_caster'] ?? new TypeCaster());
}
```

zluiten replied:
I don't like it as well, but since it was already done like this and I didn't want to distract from the essence of the changes I mimicked the style :-). Will change though!

weierophinney commented on src/Filter/ODM/AbstractFilter.php:
```diff
@@ -8,74 +8,39 @@
-                break;
+    protected function getTypeCaster()
+    {
+        if ($this->typeCaster === null) {
+            $this->typeCaster = new TypeCaster();
```
If you do as I suggest with the constructor, this becomes a moot point.

weierophinney commented on src/Filter/ORM/AbstractFilter.php:
```diff
@@ -8,19 +8,24 @@
 
+    protected $typeCaster;
+
     public function __construct($params)
     {
         $this->setFilterManager($params[0]);
+        if (isset($params[1])) {
+            $this->setTypeCaster($params[1]);
+        }
     }
```
I recognize this constructor existed before. What I'd do in this scenario is change the plugin manager to pass the following:

```php
[
    0 => $this,
    'filter_manager' => $this,
    'type_caster' => $this->resolveTypeCaster(),
]
```

and then here, you would do the following:

```php
public function __construct(array $params)
{

    $this->setFilterManager($this->extractFilterManagerFromConstructorParams($params));
    $this->setTypeCaster($params['type_caster'] ?? new TypeCaster());
}

private function extractFilterManagerFromConstructorParams(array $params)
{
    if (isset($params['filter_manager']) && $params['filter_manager'] instanceof ORMFilterManager) {
        return $params['filter_manager'];
    }
    if (isset($params[0]) && $params[0] instanceof ORMFilterManager) {
        return $params[0];
    }
    throw new InvalidArgumentException('Missing or invalid filter manager provided to ' . __CLASS__);
}
```

zluiten replied:
:+1: Will do it exactly like that

weierophinney commented on src/Filter/ORM/TypeCaster.php:
```diff
@@ -0,0 +1,70 @@
+            case 'string':
+                settype($value, 'string');
+                break;
+            case 'integer':
```
Do we need `int` here as well?

zluiten replied:
Nope, `int` is not doctrine mapping type.

weierophinney commented on src/Filter/ORM/TypeCaster.php:
```diff
@@ -0,0 +1,70 @@
+                #case 'bigint':  // Don't try to manipulate bigints?
+                settype($value, 'integer');
+                break;
+            case 'boolean':
```
Do we need `bool` here as well?

zluiten replied:
Nope, `bool` is not doctrine mapping type.

weierophinney commented on src/Filter/ORM/TypeCaster.php:
```diff
@@ -0,0 +1,70 @@
+                settype($value, 'boolean');
+                break;
+            case 'decimal':
+            case 'float':
```
Do we need `double` here as well?

zluiten replied:
Nope, `double` is not doctrine mapping type.

weierophinney commented on src/Filter/TypeCastInterface.php:
```diff
@@ -0,0 +1,14 @@
+
+interface TypeCastInterface
+{
+    public function typeCastField($metadata, $field, $value, $format = null, $doNotTypecastDatetime = false);
```
Please provide docblock annotations for each parameter. We can also bump to a PHP 7 version with this release, allowing you to add explicit typehints where applicable (e.g., `string $field`, `?string $format = null`, `bool $doNotTypecastDatetime = false`).

zluiten replied:
:+1:

weierophinney commented on src/Filter/ODM/TypeCaster.php:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        switch ($metadata->fieldMappings[$field]['type']) {
+            case 'int':
```
Do we need `integer` here?

zluiten replied:
The Doctrine ODM mapping type for integers is `integer`

weierophinney commented on src/Filter/ODM/TypeCaster.php:
```diff
@@ -0,0 +1,79 @@
+            case 'int':
+                settype($value, 'integer');
+                break;
+            case 'boolean':
```
Do we need a case for `bool` here?

zluiten replied:
The Doctrine ODM mapping type for boolean types is `boolean`

weierophinney commented on src/Filter/ODM/TypeCaster.php:
```diff
@@ -0,0 +1,79 @@
+            case 'boolean':
+                settype($value, 'boolean');
+                break;
+            case 'float':
```
Do we need a case for `double` here?

zluiten replied:
`double` is not a Doctrine ODM mapping type.

weierophinney reviewed (changes requested):
I love that you've extracted this, and I think it will add a lot of flexibility!

I've provided some notes on improvements you should make before we merge. Please bump the minimum supported PHP version to 7.2 as part of your changes (update `composer.json`, and remove job configuration from `.travis.yml`), and add typehints whereever possible on **new signatures only**.

Finally, I didn't note it, but your tests should import classes used such as `stdClass` and `DateTime`.

zluiten commented:
@weierophinney Apologies, I somehow completed missed you having reviewed this. Thanks though, I'll soon process your remarks.

zluiten commented on test/Query/Provider/DefaultOrmTest.php:
```diff
@@ -212,9 +212,11 @@ public function testGetCollectionTotal()
+
         $metadataFactory = $this->prophesize(ClassMetadataFactory::class);
-        $metadataFactory->getMetadataFor($entityClass)->willReturn($metadata);
+        $metadataFactory->getMetadataFor($entityClass)->willReturn($metadata->reveal());
```
These changes fix a faulty test with PHP 7.4.

netiul commented:
@weierophinney Thanks again for the review and apologies for the late reply. Incorporated your suggestions and meanwhile fixed a bug as noted [here](https://github.com/laminas-api-tools/api-tools-doctrine-querybuilder/pull/5#discussion_r513013983).

The Travis builds with lowest composer dependencies (`--prefer-lowest`) for 7.3 and 7.4 are (still) disabled. I'm not familiar with reason why those are disabled (I tested, builds do fail). I think it's out of scope of this PR to fix that.

zluiten commented:
@weierophinney Thanks again for the review and apologies for the late reply. Incorporated your suggestions and meanwhile fixed a bug as noted [here](https://github.com/laminas-api-tools/api-tools-doctrine-querybuilder/pull/5#discussion_r513013983).

The Travis builds with lowest composer dependencies (`--prefer-lowest`) for 7.3 and 7.4 are (still) disabled. I'm not familiar with reason why those are disabled (I tested, builds do fail). I think it's out of scope of this PR to fix that.

I've also added support for immutable date doctrine field variants. That was actually the reason I wanted to extract the type casting behavior so I could replace the TypeCaster with one that supported the immutable date types.

zluiten commented:
Hey @weierophinney, can you have another look?

weierophinney committed:
qa: fix invalid signatures due to PHPUnit update

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix phpcs configuration

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: automated CS fixes (phpcbf)

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix issues flagged by phpcs

Fixes issues flagged by phpcs that could not be autocorrected with phpcbf.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney reviewed (changes requested):
Hi, @netiul !

I merged another PR earlier today that prepares the package for PHP 8... and solves the testing issues you were observing as well (mainly by updating a few key dependencies). Since I merged that, I rebased for you... and then discovered issues in that previous patch's phpcs updates, which meant I was fixing phpcs issues... anyways, it looks good; you incorporated all the feedback I had, and we now have running tests.

The only last think I need from you: could you write-up an example for the README.md file demonstrating substituting an alternative type caster implementation? (Or is that even necessary?)

zluiten committed:
Mention custom TypeCaster implementation in the README f.e. when using custom mapping types.

Signed-off-by: Zacharias Luiten <zach@xzachly.com>

zluiten commented:
@weierophinney Thanks again. I've put something in the README.

weierophinney committed:
docs: edited docs for consistency and formatting

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
Adds CHANGELOG entry for #5

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
